### PR TITLE
Add `SqlSelect` instance for `:&`

### LINKED ISSUE #267: `SqlSelect` instances for `a :& b`
Should `a :& b` have a `SqlSelect` instance, based on that of `(a, b)`? This has come up for me when using the experimental API, since I do something like
```haskell
select $ do
  e <- from ... -- Here, the 'from' is polymorphic and user-supplied
  ... -- Using a polymorphic @e :: rec@
  pure e -- Returning @rec@
```
In the above example, where the `from` is a join, `rec` becomes `a :& b`, but since Esqueleto expects you to split the `:&` and return `(a, b)` instead, there's no appropriate instance for running the `select`.

The instance could look something like this:
```haskell
toJoin :: (a, b) -> (a :& b)
toJoin (a, b) = (a :& b)

fromJoin :: (a :& b) -> (a, b)
fromJoin (a :& b) = (a, b)

instance (SqlSelect a ra, SqlSelect b rb) => SqlSelect (a :& b) (ra :& rb) where
    sqlSelectCols esc = sqlSelectCols esc . fromJoin
    sqlSelectColCount = sqlSelectColCount . fromJoinProxy
      where
       fromJoinProxy :: Proxy (a :& b) -> Proxy (a, b)
       fromJoinProxy Proxy = Proxy
    sqlSelectProcessRow = fmap toJoin . sqlSelectProcessRow
```
This is an orphan I'm using to make my code compile, but I don't know if it's the intended way.

### PR DESCRIPTION
**Pull request by ivanbakel**
Fixes #267.

Before submitting your PR, check that you've:

- [x] Bumped the version number.
- [x] Documented new APIs with [Haddock markup](https://www.haskell.org/haddock/doc/html/index.html).
- [x] Added [`@since` declarations](http://haskell-haddock.readthedocs.io/en/latest/markup.html#since) to the Haddock.
- [x] Ran `stylish-haskell` and otherwise adhered to the [style guide](https://github.com/bitemyapp/esqueleto/blob/master/style-guide.yaml).

After submitting your PR:

- [x] Update the Changelog.md file with a link to your PR.
- [ ] Check that CI passes (or if it fails, for reasons unrelated to your change, like CI timeouts).

### CODE REVIEW

File: esqueleto.cabal [hunk 1]
```diff
@@ -1,7 +1,7 @@
 cabal-version: 1.12
 
 name:           esqueleto
-version:        3.5.0.0
+version:        3.5.1.0
```
Comment (parsonsmatt) [inline esqueleto.cabal:4]:
```suggestion
version:        3.5.1.1
```

File: changelog.md [hunk 2]
```diff
@@ -1,3 +1,10 @@
+3.5.1.0
```
Comment (parsonsmatt) [inline changelog.md:1]:
```suggestion
3.5.1.1
```

File: src/Database/Esqueleto/Experimental/From/Join.hs [hunk 3]
```diff
@@ -49,6 +49,18 @@ class ValidOnClause a
 instance {-# OVERLAPPABLE #-} ToFrom a a' => ValidOnClause a
 instance ValidOnClause (a -> SqlQuery b)
 
+-- | You may return joined values from a 'select' query - this is
+-- identical to the tuple instance, but is provided for convenience.
+--
+-- @since 3.5.1.0
```
Comment (parsonsmatt) [inline src/Database/Esqueleto/Experimental/From/Join.hs:55]:
```suggestion
-- @since 3.5.1.1
```

**Review by parsonsmatt — APPROVED** [review]
Approving conditional on merging `master` and updating the version number.

Since we're adding an instance to a class we control, I think it's fine to make this a patch version.

**Comment by ivanbakel** [comment]
>Since we're adding an instance to a class we control, I think it's fine to make this a patch version.

It's your repo, but the PVP does specify that this should be a minor version bump.

**Comment by parsonsmatt** [comment]
I'm sorry, I was mistaken about that. I'll go ahead and merge-on-green and make that change myself. 

Thanks so much for your contribution! 😄

**Comment by ivanbakel** [comment]
Since I had time, I've bumped the version to 3.5.2.

**Comment by parsonsmatt** [comment]
Thank you! I'll get this released today.

**Pull request merged**